spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -275,6 +275,10 @@ trait OptimisticTransactionImpl extends TransactionalWrite
   // Whether this transaction is creating a new table.
   private var isCreatingNewTable: Boolean = false
 
+  // Whether this transaction is overwriting the existing schema (i.e. overwriteSchema = true).
+  // When overwriting schema (and data) of a table, `isCreatingNewTable` should not be true.
```

Comment:
I agree that we can delay case 1 for a separate PR. Option 2 doesn't overwrite the schema, so this PR's optimization doesn't apply. So that leaves (3).

We should figure out whether we think that option (1) is like option (3) or not. But that doesn't need to block this PR.

Reply:
cc @ryan-johnson-databricks

Reply:
Agree. This is messy enough that we should probably scope/fix it as a separate work item. Do we have a tracking ticket so it doesn't fall through the cracks?

Reply:
Created: https://github.com/delta-io/delta/issues/1847